### PR TITLE
Guide: "Modelling the domain"

### PR DESCRIPTION
**Pull request by stephanvd**
- [x] Split getting started guides to different chapters.
- [x] Seperate tutorial of generated aggregate from adding event.
- [x] Link to concept guides when first mentioning concept.
- [ ] Start with writing an integration spec to "begin with the end in mind"

### CODE REVIEW

File: docs/docs/modelling-the-domain.md [hunk 1]
```diff
@@ -4,12 +4,114 @@ title: Modelling your Domain in Sequent
...
+{: .notice}
 
 ### The Aggregate Root
 
-In `lib/post/post.rb` we find the aggregate root.
+In `lib/post/post.rb` we find the aggregate root. This class encapsulates your business logic. Events are applied to instances of `Post` to give it its current state. We can see here that create a new `Post` will apply multiple events. Besides `PostAdded` we're also applying events to change the author, title and content. You might be tempted to group all those fields in one event, which can be a good idea if those fields always change together. We're using multiple events to emphasise that a single command does not alway correlate to a single event.
+
+```ruby
+class Post < Sequent::AggregateRoot
+  def initialize(command)
```
Comment (mattgibson) [inline docs/docs/modelling-the-domain.md]: What happens here when an aggregate is reconstructed from the event stream? Does the command somehow get passed in again automatically? Bit unclear about how this method is used later in the lifecycle when `PostAdded` has already happened and we are now handling `PostUpdated` or similar.

File: docs/docs/modelling-the-domain.md [hunk 2]
```diff
@@ -4,12 +4,114 @@ title: Modelling your Domain in Sequent
...
+  # ...
+end
+```
+
+Let's define how the domain should behave when receiving our new `PublishPost` command. Below the initialize method add:
+
+```ruby
+def publish(publication_date)
+  fail PostAlreadyPubishedError if @publication_date.any?
+  apply PostAdded
```
Comment (mattgibson) [inline docs/docs/modelling-the-domain.md]: I'm confused as to why `PostAdded` gets applied here. Is this correct? I was expecting a `PostPublished` event.

File: docs/docs/modelling-the-domain.md (hunk 1 above)
Comment (lvonk) [reply docs/docs/modelling-the-domain.md]: When reconstructing from the event stream we call rubies `allocate()` method. This "constructs" an object without calling the initialiser.

See [AggregateRoot.load_from_history](https://github.com/zilverline/sequent/blob/bf75eb86c02a3dfc2884872b55760eefaae2a7e0/lib/sequent/core/aggregate_root.rb#L46)

File: docs/docs/modelling-the-domain.md (hunk 2 above)
Comment (lvonk) [reply docs/docs/modelling-the-domain.md]: Thanks for spotting this

**Comment by mattgibson** [comment]
Hi. Thanks for a great framework! Sequent looks to be very useful for us. 

I've been starting to learn how to use it, but have found that some things in the docs are a little unclear, making it slow to get started. I've kept track of the questions that I've had in mind whilst going along, so perhaps it would be helpful when revising the docs for me to post them here:

- Is it better to have the attributes (e.g. the `title` of the `Post`) as fields on the command/event or to have a `PostAttributes` value object to encapsulate them so that the command/event has that as a single attribute instead? What are the pros and cons of each approach?
- How and when are the events applied when the aggregate is reconstructed from the event stream? The `initialize` method will be called each time a new object is instantiated, but there is logic in the examples that looks like it should only run when the initial `AddPost` command is handled.
- What does an `UpdatePost` command/event look like if I am using a HTML form to do the update (many fields)
- Why are `AggregateRoot`s and `Projector`s different classes? I know that this is the point of the CQRS pattern, but the docs and examples show no use case where the projectors do anything other than a 1-1 mapping of `Projector` to `AggregateRoot`. What would that look like?
- What is an `EventHandler`? Some examples have this, but the current docs don't. I see from the changelog that it's been renamed to `Projector`, but the examples were a source of confusion until I found this.
- What does the aggregate repository do? Does `add_aggregate` write to the DB? If so, which DB and what records are made? What does an `update_aggregate` operation look like?
- There are 3 DB tables in the `sequent_schema`: `event_records`, `command_records` and `stream_records`. What are they used for and when? Events I assume are used to rebuild the aggregates, but commands I don't see mention of being used and I don't know what stream are at all.
- Initially, it was not clear to me that the `aggregate_id` needed to be manually generated. I was assuming it would be automatically made on `repository.add_aggregate`
- Why is it not possible to do a zero downtime migration to add a new projection (leaving the existing ones alone)? Is this theoretically possible in future or will it never be?
- Is it OK to rename `aggregate_id` as `id` in the projections so as to be consistent with our other services? Does this matter or not?

Hopefully that's useful. I'm not very familiar with the CQRS pattern so perhaps some of these questions arise from seeing this project as my first experience with event sourcing. I'm happy to help with writing the documentation if that would be useful, but I don't know the answers to all of the above questions yet.

File: docs/docs/modelling-the-domain.md (hunk 1 above)
Comment (mattgibson) [reply docs/docs/modelling-the-domain.md]: Ah, I see. Never used this method before!

**Comment by lvonk** [comment]
Hi @mattgibson thanks for your detailed feedback. We are continuously improving the docs and will certainly address your points.

**Comment by lvonk** [comment]
Hi @mattgibson,

I have created a separate issue from your comments, since this PR will be merged.

See https://github.com/zilverline/sequent/issues/136

**Pull request merged**